charts/internal/cilium/charts/encryption/templates/secret.yaml

```diff
@@ -0,0 +1,34 @@
+apiVersion: v1
+kind: Secret
+metadata:
+  name: {{ .Values.global.encryption.secretName | quote }}
+  namespace: {{ .Release.Namespace }}
+type: Opaque
+data:
+  {{- $secretObj := (lookup "v1" "Secret" .Release.Namespace .Values.global.encryption.secretName) | default (dict "data" (dict)) }}
+  {{- $secretData := (get $secretObj "data") | default dict }}
+  {{- $ipsecSecret := ((get $secretData .Values.global.encryption.keyFile) | b64dec) | default "" }}
+  {{- $ipsecSecretParts := (split " " $ipsecSecret) | default (list) }}
+  {{- $ipsecKeyId := int ((trimSuffix "+" $ipsecSecretParts._0) | default .Values.global.encryption.ipsec.keyId) }}
+  {{- $ipsecEncryptionAlgorithms := $ipsecSecretParts._1 | default .Values.global.encryption.ipsec.encryptionAlgorithms }}
+  {{- $ipsecPreSharedKey := $ipsecSecretParts._2 | default .Values.global.encryption.ipsec.preSharedKey | default (randAlphaNum 32 | sha256sum) }}
+  {{- $ipsecKeySize := $ipsecSecretParts._3 | default (.Values.global.encryption.ipsec.keySize | toJson) }}
+  {{- $providedKeyId := int .Values.global.encryption.ipsec.keyId }}
+  {{- if (or (lt $providedKeyId 1) (gt $providedKeyId 15)) }}
+    {{- fail "ipsecKeyId must be a number between 1 and 15." }}
+  {{- end }}
+  {{- if (or (gt $providedKeyId $ipsecKeyId) (and (eq $ipsecKeyId 15) (eq $providedKeyId 1))) }}
+    {{- if eq $ipsecPreSharedKey .Values.global.encryption.ipsec.preSharedKey }}
+      {{- fail "it makes no sense to increase the KeyId without changing the ipsecPreSharedKey." }}
+    {{- end }}
+    {{- $ipsecKeyId = $providedKeyId }}
+    {{- $ipsecPreSharedKey = .Values.global.encryption.ipsec.preSharedKey | default (randAlphaNum 32 | sha256sum) }}
+  {{- else if (and (ne .Values.global.encryption.ipsec.preSharedKey "") (ne $ipsecPreSharedKey .Values.global.encryption.ipsec.preSharedKey)) }}
+    {{- if (eq $ipsecKeyId 15) }}
+      {{- $ipsecKeyId = 0 }}
+    {{- end }}
+    {{- $ipsecKeyId = add $ipsecKeyId 1 }}
+    {{- $ipsecPreSharedKey = .Values.global.encryption.ipsec.preSharedKey }}
+  {{- end }}
+  # per-tunnel keys will be derived from the secret by adding the `+` sign to the keyId
+  {{ .Values.global.encryption.keyFile }}: {{ printf "%s+ %s %s %s" ($ipsecKeyId | toJson) $ipsecEncryptionAlgorithms $ipsecPreSharedKey $ipsecKeySize | b64enc | quote }}
```

Comment:
There is quiet some logic implemented in this helm template.
Wouldn't it make sense to implement some parts in go?

Did you think about credential rotation?

Reply:
Did not think about it yet. The logic was implemented because of keeping generated keys and to allow easy credential rotation (increment keyId or provide a new preSharedKey).
If it's more readable, we should go for the go implementation.

Reply:
+1 on moving this to Go. Logic in helm template is always very hard to debug.

Reply:
There is this credential rotation tool in the cilium cli: https://github.com/cilium/cilium/blob/main/cilium-cli/encrypt/ipsec_rotate_key.go
We should think about reusing some of this code. The question would be how to trigger rotation?

Reply:
@hown3d incrementing keyId or providing a new preSharedKey could be a trigger

Reply:
Changes of preSharedKey is definitely a trigger.
Maybe we could hook into the credential rotation from gardener? I believe that's the most reasonable way for users to trigger it.